org.jcryptool.visual.rabin/src/org/jcryptool/visual/rabin/HandleCryptosystemTextbook.java

```diff
@@ -0,0 +1,974 @@
+			return;
+		
+		countClicksForPlaintexts[idxToSet] = (countClicksForPlaintexts[idxToSet] + 1) % 2;
+		
+		if(countClicksForPlaintexts[idxToSet] == 1) {
+			clickedPlaintexts.get(idx).set(idxToSet, true);
+			String key = String.valueOf(idx) + String.valueOf(idxToSet);
+			String val = plaintexts.get(idx).get(idxToSet);
+			currentSelectedPlaintexts.put(key, val);
+			String currentSelectedPlaintextsAsStr = getCurrentSelectedPlaintexts();
+			cstb.txtChosenPlaintexts.setText(currentSelectedPlaintextsAsStr);
+			setColorPlaintext(cstb, idxToSet, colorToSetBG, colorToSetFG);
+		}
+		
+		if(countClicksForPlaintexts[idxToSet] == 0) {
+			clickedPlaintexts.get(idx).set(idxToSet, false);
+			String key = String.valueOf(idx) + String.valueOf(idxToSet);
+			currentSelectedPlaintexts.remove(key);
+			String currentSelectedPlaintextsAsStr = getCurrentSelectedPlaintexts();
+			cstb.txtChosenPlaintexts.setText(currentSelectedPlaintextsAsStr);
+			setColorPlaintext(cstb, idxToSet, colorToResetBG, colorToResetFG);
+		}
+	}
+	
+	
+	
+	public void txtFirstPlaintextMouseDownAction(CryptosystemTextbookComposite cstb) {
+		if(clickedPlaintexts == null)
+			return;
+		
+		int idxToSet = 0;
+		txtPlaintextMouseDownAction(idxToSet, cstb, GUIHandler.colorSelectControlBG, GUIHandler.colorDeselectControlBG, GUIHandler.colorSelectControlFG, GUIHandler.colorDeselectControlFG);
+	}
+	
+	
+	public void txtSecondPlaintextMouseDownAction(CryptosystemTextbookComposite cstb) {
+		if(clickedPlaintexts == null)
+			return;
+		
+		int idxToSet = 1;
+		txtPlaintextMouseDownAction(idxToSet, cstb, GUIHandler.colorSelectControlBG, GUIHandler.colorDeselectControlBG, GUIHandler.colorSelectControlFG, GUIHandler.colorDeselectControlFG);
+	}
+	
+	
+	public void txtThirdPlaintextMouseDownAction(CryptosystemTextbookComposite cstb) {
+		if(clickedPlaintexts == null)
+			return;
+		
+		int idxToSet = 2;
+		txtPlaintextMouseDownAction(idxToSet, cstb, GUIHandler.colorSelectControlBG, GUIHandler.colorDeselectControlBG, GUIHandler.colorSelectControlFG, GUIHandler.colorDeselectControlFG);
+	}
+	
+	
+	
+	public void txtFourthPlaintextMouseDownAction(CryptosystemTextbookComposite cstb) {
+		if(clickedPlaintexts == null)
+			return;
+		
+		int idxToSet = 3;
+		txtPlaintextMouseDownAction(idxToSet, cstb, GUIHandler.colorSelectControlBG, GUIHandler.colorDeselectControlBG, GUIHandler.colorSelectControlFG, GUIHandler.colorDeselectControlFG);
+	}
+	
+	
+	
+	public void btnWriteToJCTeditor(CryptosystemTextbookComposite cstb) throws PartInitException {
+		String ciphertext = cstb.txtCiphertext.getText();
+		int len = ciphertext.length();
+		String ciphertextForFile= ""; //$NON-NLS-1$
+		int lineLimit = 100;
+		
+		if(len < 100) {
+			ciphertextForFile = ciphertext;
+		}
+		else {
+		
+			for(int i = 0; i < len; i++) {
+				ciphertextForFile += ciphertext.charAt(i);
+				if((i+1) % lineLimit == 0)
+					ciphertextForFile += "\n"; //$NON-NLS-1$
+			}
```

Comment:
Iterating over characters and using the `+` operator within a loop is quite inefficient, as it has to manage a lot of pointers on the heap. Rather use a `StringBuilder` in Java.
Since we want to iterate over chunks / substrings, we can use those methods:

```suggestion
		StringBuilder sb = new StringBuilder();
		for (int chunk = 0; chunk < ciphertext.length(); chunk += lineLimit) {
		        int chunkEnd = Math.min(chunk + lineLimit, cipherText.length());
			sb.append(cipherText.subSequence(chunk, chunkEnd));
			sb.append('\n');
		}
```